### ISSUE TITLE
Error running lstchain_mc_sensitivity.py script

### ISSUE THREAD

**Issue by andres-baquero** [cta-observatory/cta-lstchain#299]
I'm using lstchain v0.4.4 and I've installed ctaplot.

I'm getting this error running lstchain_mc_sensitivity.py:

`AttributeErrorTraceback (most recent call last)
<ipython-input-11-afaf184bc95e> in <module>
      6                                                                               n_bins_energy, n_bins_gammaness,
      7                                                                               n_bins_theta2, noff,
----> 8                                                                               obstime)

/fefs/home/andres.baquero/lstchain-dev/lstchain_v044/cta-lstchain/lstchain/mc/sensitivity.py in find_best_cuts_sensitivity(simtelfile_gammas, simtelfile_protons, dl2_file_g, dl2_file_p, nfiles_gammas, nfiles_protons, n_bins_energy, n_bins_gammaness, n_bins_theta2, noff, obstime)
    332 
    333     # Pass units to GeV and cm2
--> 334     mc_par_g['emin'] = mc_par_g['emin'].to(u.GeV)
    335     mc_par_g['emax'] = mc_par_g['emax'].to(u.GeV)
    336 

AttributeError: 'numpy.float64' object has no attribute 'to'
`
Thanks for your help

**Comment by rlopezcoto** [comment]
Hi @andres-baquero, the minimum energy (`mc_par_g['emin']`) and the maximum energy (`mc_par_g['emax']`) should be read from the original MC file (`simtelfile_gammas`), which one is the one you are using?

**Comment by andres-baquero** [comment]
Hi @rlopezcoto. I'm using a dl1 file of gamma point merged with all runs produced from:
` /fefs/aswg/data/mc/DL0/20190415/gamma/south_pointing`

**Comment by rlopezcoto** [comment]
Those are DL0 according to the directory, right?

**Comment by andres-baquero** [comment]
Yes. I'm using DL1 files for protons and gammas that have been produced from  this DL0 files.

**Comment by rlopezcoto** [comment]
Can you give me the exact command you are using (with all the input files) so I can try to reproduce the error? 
Thanks

**Comment by andres-baquero** [comment]
I'm not sure if is related with the error. Looking in the code I found that `find_best_cuts_sensitivity` receive this parameters:
 
```
def find_best_cuts_sensitivity(simtelfile_gammas, simtelfile_protons,
         dl2_file_g, dl2_file_p,
         nfiles_gammas, nfiles_protons,
         n_bins_energy, n_bins_gammaness, n_bins_theta2, noff,
         obstime = 50 * 3600 * u.s):
```


But the script is sending instead of `nfiles_gammas` and `nfiles_protons` the number of telescopes:

```


 ntelescopes_gamma = 4
    ntelescopes_protons = 4
    n_bins_energy = 20  #  Number of energy bins
    n_bins_gammaness = 11  #  Number of gammaness bins
    n_bins_theta2 = 10  #  Number of theta2 bins
    obstime = 50 * 3600 * u.s
    noff = 5




energy, best_sens, result, units, gcut, tcut = find_best_cuts_sensitivity(args.dl1file_gammas,
                                                                              args.dl1file_protons,
                                                                              args.dl2_file_g_cuts,
                                                                              args.dl2_file_p_cuts,
                                                                              ntelescopes_gamma, ntelescopes_protons,
                                                                              n_bins_energy, n_bins_gammaness,
                                                                              n_bins_theta2, noff,
                                                                              obstime)
```

**Comment by moralejo** [comment]
> /fefs/home/andres.baquero/lstchain-dev/lstchain_v044/cta-lstchain/lstchain/mc/sensitivity.py in find_best_cuts_sensitivity(simtelfile_gammas, simtelfile_protons, dl2_file_g, dl2_file_p, nfiles_gammas, nfiles_protons, n_bins_energy, n_bins_gammaness, n_bins_theta2, noff, obstime)
> 332
> 333 # Pass units to GeV and cm2
> --> 334 mc_par_g['emin'] = mc_par_g['emin'].to(u.GeV)
> 335 mc_par_g['emax'] = mc_par_g['emax'].to(u.GeV)
> 336
> 
> AttributeError: 'numpy.float64' object has no attribute 'to'
> 
Did that ever work? Seems to me the values mc_par_g['emin'] are those directly read from the hdf5 file, so they have no associated units. Hence we should do mc_par_g['emin']*u.TeV  (if the _number_ stored in the file is in TeV), and then do the conversion to GeV.

**Comment by andres-baquero** [comment]
Hi @moralejo yes that was the problem in this case.

Another bug is that in sensitivity.py/find_best_cuts_sensitivity

`# Pass units to GeV and cm2
    mc_par_g['emin'] = mc_par_g['emin'].to(u.GeV)
    mc_par_g['emax'] = mc_par_g['emax'].to(u.GeV)

    mc_par_p['emin'] = mc_par_p['emin'].to(u.GeV)
    mc_par_p['emax'] = mc_par_p['emax'].to(u.GeV)
`
Are duplicated.

**Comment by vuillaut** [comment]
Hi @andres-baquero 
If you fixed the bug, can you make a PR, please?
ps: please get everything in TeV as the default unit.

Thanks.

**Comment by andres-baquero** [comment]
Hi @vuillaut . Ok I will make the PR.

Just one question @rlopezcoto the script receive 4 DL2 files:

`--gammadl2-cuts DL2_FILE_G_CUTS, 
  --protondl2-cuts DL2_FILE_P_CUTS, 
  --gammadl2-sens DL2_FILE_G_SENS, 
  --protondl2-sens DL2_FILE_P_SENS
`

Which is the difference between this DL2 files ?.

**Comment by SeiyaNozaki** [comment]
Hi @andres-baquero ,
I reproduced the issue like below;

```
>>> from lstchain.mc.sensitivity import process_mc
>>> dl1_gamma="/fefs/aswg/data/mc/DL1/20190415/gamma/south_pointing/20191128_v.0.3.1_v00/dl1_gamma_south_pointing_20191128_v.0.3.1_v00_DL1_testing.h5"
>>> dl2_gamma="/fefs/aswg/data/mc/DL2/20190415/gamma/south_pointing/20191128_v.0.3.1_v00/dl2_dl1_gamma_south_pointing_20191128_v.0.3.1_v00_DL1_testing.h5"
>>> gammaness_g, theta2_g, e_reco_g, e_true_g, mc_par_g, events_g = process_mc(dl1_gamma, dl2_gamma, 'gamma')
Table '/simulation/run_config' has column 'obs_id' that is not in container MCHeaderContainer. It will be skipped
Table '/simulation/run_config' is missing column 'prod_site_array' that is in container MCHeaderContainer. It will be skipped.
Table '/simulation/run_config' is missing column 'prod_site_coord' that is in container MCHeaderContainer. It will be skipped.
Table '/simulation/run_config' is missing column 'prod_site_subarray' that is in container MCHeaderContainer. It will be skipped.
>>> mc_par_g['emin']
0.004999999888241291
```

But with DL1 file which I generated by using the latest lstchain, mc_par_g['emin'] has TeV unit.
I'm checking...

```
>>> dl1_gamma="/fefs/aswg/workspace/seiya.nozaki/data/MC/IC6_3/DL1/gamma_point/dl1_gamma_20deg_0deg_run1___cta-prod3-demo-2147m-LaPalma-baseline-mono_off0.4.simtel.h5"
>>> gammaness_g, theta2_g, e_reco_g, e_true_g, mc_par_g, events_g = process_mc(dl1_gamma, dl2_gamma, 'gamma')
Table '/simulation/run_config' has column 'obs_id' that is not in container MCHeaderContainer. It will be skipped
Table '/simulation/run_config' is missing column 'prod_site_array' that is in container MCHeaderContainer. It will be skipped.
Table '/simulation/run_config' is missing column 'prod_site_coord' that is in container MCHeaderContainer. It will be skipped.
Table '/simulation/run_config' is missing column 'prod_site_subarray' that is in container MCHeaderContainer. It will be skipped.
>>> mc_par_g['emin']
<Quantity 0.005 TeV>
```

**Comment by vuillaut** [comment]
Actually I am puzzled with the bug here.
Checking the code, `mc_par_g['emin']` must have unit here.
It uses `read_simu_info_hdf5` that returns a `MCHeaderContainer()` with TeV unit for `energy_range_min`. And this part of the code is acutally unit tested `test_read_sim_par`.

**Comment by andres-baquero** [comment]
So @vuillaut the PR is still needed ?.

**Comment by vuillaut** [comment]
> So @vuillaut the PR is still needed ?.

If I am correct, actually not.

**Comment by vuillaut** [comment]
Hi @andres-baquero 
I finally found the issue. It actually comes from the merging, which is done in a generic way for all tables, loosing the unit information. Nasty one :-/

**Comment by rlopezcoto** [comment]
Sorry for arriving so late at the discussion.
@moralejo: the hdf5 can have units, but as reported by @vuillaut, they were stripped away in the merging.
@andres-baquero: there are not 4 DL2 files, there are two sets of DL2 files, the ones to calculate the optimal theta2 and gammaness cuts (`gammadl2-cuts` `protondl2-cuts`) and the ones to apply them to calculate the sensitivity (`gammadl2-sens` `protondl2-sens`). This was set like this by @misabelber not to have a biased sensitivity.
@SeiyaNozaki: thanks for the checks!
@vuillaut: thanks a lot for spotting the problem, is #300 working it out or do we need another fix?

**Comment by rlopezcoto** [comment]
Fixed in #299
